<html>
<!--
 * Jalview - A Sequence Alignment Editor and Viewer (2.11.2.5)
 * Copyright (C) 2022 The Jalview Authors
 * 
 * This file is part of Jalview.
 * 
 * Jalview is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License 
 * as published by the Free Software Foundation, either version 3
 * of the License, or (at your option) any later version.
 *  
 * Jalview is distributed in the hope that it will be useful, but 
 * WITHOUT ANY WARRANTY; without even the implied warranty 
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR 
 * PURPOSE.  See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Jalview.  If not, see <http://www.gnu.org/licenses/>.
 * The Jalview Authors are detailed in the 'AUTHORS' file.
 -->
<head>
<title>Split Frame Views</title>
</head>
<body>
  <p>
    <strong>Split Frame Views</strong>
  </p>
  <p>
    Jalview provides a special viewing mode to show Coding DNA (cDNA)
    and protein product alignments as a split view, with cDNA above and
    protein below. The two alignments are linked, allowing editing and
    analysis to be performed at both the peptide and nucleotide level.
    Linked protein alignments also have an additional <strong>cDNA
      Consensus</strong> annotation row, showing the distribution of codons at
    each column of the protein alignment.
  </p>
  <p>
    Split Frame views can be <a href="#opensplit">created in a
      number of ways</a>. In the Jalview Desktop, Split Frame views are
    saved in Jalview Projects, like any other alignment view.
  </p>
  <p><strong><a name="virtualfeats">Virtual Sequence Features</a></strong>
  </p>
  <p>Since Jalview 2.11.1, sequence features displayed in either CDS
    or Protein can be shown as 'Virtual Features' mapped onto the other
    view. Features describing non-synonymous genomic sequence variation
    in coding regions are dynamically translated when shown in the
    protein view.</p>
  <p>The display of virtual features and whether they are rendered
    below or above features local to the aligned sequences is controlled
    by a check box shown in the view's Sequence Feature Settings dialog
    box. For convenience, Protein and CDS feature settings for a linked
    CDS/Protein view are shown as tabs in a single dialog box, allowing
    easy access to display and filter settings for each view.</p>
  <p>
    <img width="631" align="center"
      alt="Linked CDS and Protein Feature Settings showing Virtual Feature checkbox"
      src="linkedfeatsettings.png" /> <em>CDS and Protein feature
      settings tabs and corresponding views showing 'Virtual Features'
      from each view overlaid on the other (created with Jalview 2.11.1.0).</em>
  </p>
  <p>
    When virtual features are enabled, they are also shown on any linked
    3D structure views when 'Colour by Sequence' is enabled, and
    exported as GFF and Jalview Features files (mapped to their
    associated virtual coordinates). Both the original and the mapped
    locations are also included in <a href="seqfeaturereport.html">Sequence
      Feature Reports</a>.
  </p>
  <p>
    <strong>Operations supported in Split Frame Mode</strong>
  </p>
  <p>Split Frame views allow the following:
  <ul>
    <li>Mouseover or scrolling of either alignment is followed by
      the other (unless you turn off <strong><a
        href="../menus/alwview.html">"View&#8594;Automatic
          Scrolling"</a></strong>)
    </li>
    <li>On selecting rows, columns or regions in one alignment, the
      corresponding selection is made in the other</li>
    <li>Sequence ordering in one alignment (using the cursor, or <strong><a
        href="../calculations/sorting.html">"Calculate&#8594;Sort")</a></strong>
      is also applied to the other
    </li>
    <li>Editing (gap insertion / deletion) in the protein alignment
      is reflected in the cDNA (but not vice versa)</li>
    <li>Any trees imported or created with <strong><a
        href="../calculations/tree.html">"Calculate Tree"</a></strong> on
      one of the views allow both cDNA and Protein views to be grouped,
      coloured or sorted.
    </li>
    <li>To allow for the different widths in cDNA and Protein
      alignments, the <strong><a href="../menus/alwformat.html">"Format&#8594;Font"</a></strong>
      menu option has an option 'Scale protein residues to codons'. This
      option will make each protein residue the same width as a DNA
      codon (so the alignments 'line up' vertically).</li>
    <li><a name="mirror" />The 'Use same font for cDNA and peptide'
      checkbox, when enabled, ensures that font or font-size changes in
      either the cDNA or Protein alignment will also be mirrored. (<em>Added
        in 2.10.2</em>)</li>
    <li><strong>"View&#8594;Protein"</strong> (in the cDNA panel)
      or <strong>"View&#8594;Nucleotide"</strong> (in the protein panel)
      allows you to show or hide one or other of the linked alignment
      panels.</li>
    <li>Panel heights are adjusted by dragging the divider between
      them using the mouse</li>
    <li><a href="../menus/alwview.html"><strong>"View&#8594;New
          View / Expand Views / Gather Views"</strong></a> behave as for a normal
      alignment window, but always create new views as Split Frames</li>
  </ul>
  <p>
    An alignment annotation row on the protein alignment shows the <strong><a
      href="../calculations/consensus.html">cDNA consensus</a></strong> for
    each peptide column.<br /> This consensus may reveal variation in
    nucleotides coding for conserved protein residues.
  </p>

  <a name="opensplit" />
  <p>
    <strong>Opening a Split Frame View</strong>
  </p>
  <p>A Split Frame View can be opened in one of the following ways:</p>
  <p>
    <strong><em>Add Sequences</em></strong>
  </p>
  <p>
    If you add (coding) DNA sequences to an open peptide alignment, or
    vice versa, <em>and</em> at least one DNA sequence translates to one
    of the peptide sequences, then the option to open in a split window
    is offered. The DNA may include start and/or stop codons, but no
    non-coding (intron) sequence.<br> If more than one cDNA variant
    is present in the alignment, Jalview will first try to match these
    to protein sequences based on any retrieved cross-references, and
    failing that, pairwise as they are ordered in the alignments.
  <p>This option is available in Jalview Desktop (when adding
    sequences by any supported method), and Jalview applet (adding from
    textbox). The additional options below apply to Jalview Desktop
    only.</p>

  <p>
    <strong><em>Translate as cDNA</em></strong>
  </p>
  <p>
    Menu option <strong><a href="../menus/alwcalculate.html">"Calculate&#8594;Translate
        as cDNA"</a></strong> is available for a nucleotide alignment. Selecting this
    option shows the DNA and its calculated protein product in a Split
    Frame view.
  </p>

  <p>
    <strong><em>Get Cross-References</em></strong>
  </p>
  <p>
    Menu option <strong><a href="../menus/alwcalculate.html">"Calculate&#8594;Get
        Cross-References"</a></strong> is available for fetched sequences which have
    cross-references to other databases. On selecting protein
    cross-references (for a cDNA alignment), or DNA xrefs (for peptide),
    a Split Frame view is opened showing cDNA and peptide.
  </p>

  <p>
    <strong><em>Realign a Split View</em></strong>
  </p>
  <p>If you invoke a web service to realign either half of a Split
    Frame, then the resulting realignment is displayed in a new Split
    Frame.</p>
  <ul>
    <li>the alignment you chose to realign (for example, peptide)
      is displayed as aligned by the external web service</li>
    <li>Jalview reconstructs the alignment of its complement (in
      this case, cDNA) by inserting gaps in the corresponding positions</li>
  </ul>
  <p>
    <a name="reconalignment" /><strong>Reconstructed
      Alignments</strong>
  </p>
  <p>
    Reconstructed alignments are typically <em>not</em> the same as the
    alignment produced by aligning the complement sequence set directly
    with the external service. However, in the case of protein
    alignments, a reconstructed cDNA alignment is often more reliable
    than one calculated without coding information. Reconstructed cDNA
    alignments are also more informative than the original protein
    alignment when calculating phylogenetic trees or performing other
    kinds of molecular evolution analysis.
  </p>
  <p>
    <em>Split Frame Views were introduced in Jalview 2.9</em>
  </p>
</body>
</html>
