<html>
<!--
 * Jalview - A Sequence Alignment Editor and Viewer (2.11.2.5)
 * Copyright (C) 2022 The Jalview Authors
 * 
 * This file is part of Jalview.
 * 
 * Jalview is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License 
 * as published by the Free Software Foundation, either version 3
 * of the License, or (at your option) any later version.
 *  
 * Jalview is distributed in the hope that it will be useful, but 
 * WITHOUT ANY WARRANTY; without even the implied warranty 
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR 
 * PURPOSE.  See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Jalview.  If not, see <http://www.gnu.org/licenses/>.
 * The Jalview Authors are detailed in the 'AUTHORS' file.
 -->
<head>
<title>UniProtKB query fields</title>
</head>

<body>
  <p>
    <strong>UniProtKB query fields</strong>
  </p>
  <p>
    Supported query fields for searching specific data in UniProtKB (see
    also <a href="uniprotsequencefetcher.html#text-search">query
      syntax</a>).
  </p>

<table border="1" width="95%">
<thead>
<tr>
<th>rest.uniprot.org field</th>
<th>rest.uniprot.org example</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>accession</td>
<td>accession:P62988</td>
<td>The old behaviour was to list all entries with primary or secondary accession number P62988. The new behaviour will list all primary / canonical isoform accessions P62988. To search over secondary accessions, we have introduced the <code>sec_acc</code> field.</td>
</tr>
<tr>
<td>active</td>
<td>active:false</td>
<td>Lists all obsolete entries.</td>
</tr>
<tr>
<td>Refer to the page: Sequence Annotations</td>
<td></td>
<td>Lists all entries with:<ol><li>any general annotation (comments [CC])</li> <li>any sequence annotation (features [FT])</li><li>at least one amino acid modified with a Pyrrolidone carboxylic acid group</li></ol></td>
</tr>
<tr>
<td>lit_author</td>
<td>lit_author:ashburner</td>
<td>Lists all entries with at least one reference co-authored by Michael Ashburner.</td>
</tr>
<tr>
<td>protein_name</td>
<td>protein_name:CD233</td>
<td>Lists all entries whose cluster of differentiation number is CD233 (see cdlist.txt).</td>
</tr>
<tr>
<td>chebi</td>
<td>chebi:18420</td>
<td>Lists all entries which are associated with the small molecule corresponding to ChEBI identifier 18420, Mg(2+) (see How can I search UniProt for chemical or reaction data?).</td>
</tr>
<tr>
<td>uniprot_id (/uniref), then uniref_cluster_90 (/uniprotkb)</td>
<td><ol><li>uniprot_id:A5YMT3 to find cluster UniRef90_P00395</li><li>uniref_cluster_90:UniRef90_P00395</li></ol></td>
<td>Find all entries in the UniRef 90% identity cluster whose representative sequence is UniProtKB entry A5YMT3 (about UniRef).</td>
</tr>
<tr>
<td>xref<em>count_pdb (or xref_count</em>)</td>
<td>xref_count_pdb:[20 TO *]</td>
<td>Lists all entries with 20 or more cross-references to PDB</td>
</tr>
<tr>
<td>date_created</td>
<td>date_created:[2012-10-01 TO *]</td>
<td>Lists all entries created since October 1st 2012.</td>
</tr>
<tr>
<td>database, xref</td>
<td><ol><li>database:pfam</li><li>xref:pdb-1aut</li></ol></td>
<td>Lists all entries with:<ol><li>a cross-reference to the Pfam database</li><li>a cross-reference to the PDB database entry 1aut</li></ol> (see Databases cross-referenced in UniProtKB and Database mapping)</td>
</tr>
<tr>
<td>ec</td>
<td>ec:3.2.1.23</td>
<td>Lists all beta-galactosidases (Enzyme nomenclature database).</td>
</tr>
<tr>
<td>Refer to the pages: Comments or Sequence Annotations</td>
<td></td>
<td>Lists all entries with: <ol><li>a signal sequence whose positions have been experimentally proven</li> <li>experimentally proven phosphoserine sites</li> <li>a function manually asserted according to rules</li></ol>(see Evidence attribution)</td>
</tr>
<tr>
<td>existence</td>
<td>existence:3</td>
<td>See Protein existence criteria.</td>
</tr>
<tr>
<td>family</td>
<td>family:serpin</td>
<td>Lists all entries belonging to the Serpin family of proteins (Index of protein domains and families).</td>
</tr>
<tr>
<td>fragment</td>
<td>fragment:true</td>
<td>Lists all entries with an incomplete sequence.</td>
</tr>
<tr>
<td>gene</td>
<td>gene:HPSE</td>
<td>Lists all entries for proteins encoded by gene HPSE, but also by HPSE2.</td>
</tr>
<tr>
<td>gene_exact</td>
<td>gene_exact:HPSE</td>
<td>Lists all entries for proteins encoded by gene HPSE, but excluding variations like HPSE2 or HPSE_0.</td>
</tr>
<tr>
<td>go</td>
<td>go:0015629)</td>
<td>Lists all entries associated with the GO term Actin cytoskeleton and any subclasses</td>
</tr>
<tr>
<td>virus_host_name, virus_host_id</td>
<td>virus_host_id:10090</td>
<td>Lists all entries for viruses infecting Mus musculus (Mouse)</td>
</tr>
<tr>
<td>accession_id</td>
<td>accession_id:P00750</td>
<td>Returns the entry with the primary accession number P00750.</td>
</tr>
<tr>
<td>inchikey</td>
<td>inchikey:WQZGKKKJIJFFOK-GASJEMHNSA-N</td>
<td>Returns entries associated with the small molecule identified by the InChIKey WQZGKKKJIJFFOK-GASJEMHNSA-N, i.e. D-glucopyranose (see How can I search UniProt for chemical or reaction data?). To get the CHEBI identifier for an Inchikey value, one can now use the advanced search builder.</td>
</tr>
<tr>
<td>protein_name</td>
<td>protein_name:Anakinra</td>
<td>Lists all entries whose protein name includes the "International Nonproprietary Name" is Anakinra.</td>
</tr>
<tr>
<td>interactor</td>
<td>interactor:P00520</td>
<td>Lists all entries describing interactions with the protein described by entry P00520.</td>
</tr>
<tr>
<td>keyword</td>
<td><ol><li>keyword:toxin</li><li>keyword:KW-0800</li></ol></td>
<td><ol><li>Lists all entries associated with a keyword matching "Toxin" in its name or description (UniProtKB Keywords).</li> <li>Lists all entries associated with the UniProtKB keyword Toxin.</li></ol></td>
</tr>
<tr>
<td>length</td>
<td>length:[500 TO 700]</td>
<td>Lists all entries describing sequences of length between 500 and 700 residues.</td>
</tr>
<tr>
<td>mass</td>
<td>mass:[500000 TO *]</td>
<td>Lists all entries describing sequences with a mass of at least 500,000 Da.</td>
</tr>
<tr>
<td>cc_mass_spectrometry</td>
<td>cc_mass_spectrometry:maldi</td>
<td>Lists all entries for proteins identified by: matrix-assisted laser desorption/ionization (MALDI), crystallography (X-Ray). The <code>method</code> field searches names of physico-chemical identification methods in the 'Biophysicochemical properties' subsection of the 'Function' section, the 'Publications' and 'Cross-references' sections.</td>
</tr>
<tr>
<td>date_modified</td>
<td>modified:[2012-01-01 TO 2019-03-01] AND active:true</td>
<td>Lists all active entries that were last modified between January and March 2019.</td>
</tr>
<tr>
<td>protein_name</td>
<td>protein_name:"prion protein"</td>
<td>Lists all entries for prion proteins.</td>
</tr>
<tr>
<td>organelle</td>
<td>organelle:Mitochondrion</td>
<td>Lists all entries for proteins encoded by a gene of the mitochondrial chromosome.</td>
</tr>
<tr>
<td>organism_name, organism_id</td>
<td><ol><li>organism_name:"Ovis aries"</li><li>organism_id:9940</li><li>organism_name:sheep</li></ol></td>
<td>Lists all entries for proteins expressed in sheep (first 2 examples) and organisms whose name contains the term "sheep" (UniProt taxonomy).</td>
</tr>
<tr>
<td>plasmid</td>
<td>plasmid:ColE1</td>
<td>Lists all entries for proteins encoded by a gene of plasmid ColE1 (Controlled vocabulary of plasmids).</td>
</tr>
<tr>
<td>proteome</td>
<td>proteome:UP000005640</td>
<td>Lists all entries from the human proteome.</td>
</tr>
<tr>
<td>proteomecomponent</td>
<td>proteomecomponent:"chromosome 1" AND organism_id:9606</td>
<td>Lists all entries from the human chromosome 1.</td>
</tr>
<tr>
<td>sec_acc</td>
<td>sec_acc:P02023</td>
<td>Lists all entries that were created from a merge with entry P02023 (see FAQ).</td>
</tr>
<tr>
<td>reviewed</td>
<td>reviewed:true</td>
<td>Lists all UniProtKB/Swiss-Prot entries (about UniProtKB).</td>
</tr>
<tr>
<td>scope</td>
<td>scope:mutagenesis</td>
<td>Lists all entries containing a reference that was used to gather information about mutagenesis (Entry view: "Cited for", See 'Publications' section of the user manual).</td>
</tr>
<tr>
<td>sec_acc</td>
<td>sec_acc:P62988</td>
<td>Lists all entries containing a secondary accession P62988.</td>
</tr>
<tr>
<td>sequence</td>
<td>accession:P05067-9 AND is_isoform:true</td>
<td>Lists all entries containing a link to isoform 9 of the sequence described in entry P05067. Allows searching by specific sequence identifier.</td>
</tr>
<tr>
<td>date_sequence_modified</td>
<td><ol><li>date_sequence_modified:[2012-01-01 TO 2012-03-01]</li><li>date_sequence_modified:[2012-01-01 TO 2012-03-01]</li></ol></td>
<td><ol><li>Lists all entries whose sequences were last modified between January and March 2012.</li><li>Lists all UniProtKB/Swiss-Prot entries whose sequences were modified after the start of 2012.</li></ol></td>
</tr>
<tr>
<td>strain</td>
<td>strain:wistar</td>
<td>Lists all entries containing a reference relevant to strain wistar (Lists of strains in reference comments and Taxonomy help: organism strains).</td>
</tr>
<tr>
<td>taxonomy_name, taxonomy_id</td>
<td><ol><li>taxonomy_name:mammal</li><li>taxonomy_id:40674</li></ol></td>
<td>Lists all entries for proteins expressed in Mammals. This field is used to retrieve entries for all organisms classified below a given taxonomic node (taxonomy classification).</td>
</tr>
<tr>
<td>tissue</td>
<td>tissue:liver</td>
<td>Lists all entries containing a reference describing the protein sequence obtained from a clone isolated from liver (Controlled vocabulary of tissues).</td>
</tr>
<tr>
<td>cc_webresource</td>
<td>cc_webresource:wikipedia</td>
<td>Lists all entries for proteins that are described in Wikipedia.</td>
</tr>
</tbody>
</table>


</body>
</html>
