<html>
<!--
 * Jalview - A Sequence Alignment Editor and Viewer (2.11.2.5)
 * Copyright (C) 2022 The Jalview Authors
 * 
 * This file is part of Jalview.
 * 
 * Jalview is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License 
 * as published by the Free Software Foundation, either version 3
 * of the License, or (at your option) any later version.
 *  
 * Jalview is distributed in the hope that it will be useful, but 
 * WITHOUT ANY WARRANTY; without even the implied warranty 
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR 
 * PURPOSE.  See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Jalview.  If not, see <http://www.gnu.org/licenses/>.
 * The Jalview Authors are detailed in the 'AUTHORS' file.
 -->
<head>
<title>PDB Viewing</title>
</head>
<body>
  <p>
    <strong>Discovering and Viewing PDB and 3D-Beacons structures</strong>
  </p>
  Jalview can be used to explore the 3D structures of sequences in an
  alignment by following the steps below:
  <ol>
    <li>Select the <strong>"3D Structure Data..."</strong> option
      from a sequence's <a href="../menus/popupMenu.html">pop-up
        menu</a> to open the <a href="structurechooser.html">Structure
        Chooser</a> dialog box.
      <ul>
        <li>If one or more structures exist in the PDB for the given
          sequence, the <a href="structurechooser.html">Structure
            Chooser</a> dialog will open with them listed in the results
          pane.
        </li>
        <li>However, if no structure was found, the <a
          href="structurechooser.html">Structure Chooser</a> interface
          will present options for manual association of PDB structures.
        </li>
	<li>
          Since Jalview 2.11.2 you can also search the 3D-Beacons Network
	  for structures.  The 3D-Beacons Network acts as a one-stop shop for protein
	  structures by combining and standardising data from several providers.
	  See <a href="structurechooser.html#3dbeaconssearch">3D-Beacons Search</a> for
	  instructions.
	</li>
      </ul>
    </li>
    <li><strong>Selecting PDB Structures</strong><br />
      If one or more structures have been found from the PDB, you can select
      the structures that you want to open and view by selecting them
      with the mouse and keyboard.<br />By default, if structures were
      discovered, then some will already be selected according to the
      criteria shown in the drop-down menu. The default criteria is
      'highest resolution', simply choose another to pick structures in
      a different way.<br />
      <ul>
        <li><strong>Viewing Cached Structures</strong><br />If
          previously downloaded structures are available for your
          sequences, the structure chooser will automatically offer them
          via the <strong>Cached Structures</strong> view. If you wish
          to download new structures, select one of the PDBe selection
          criteria from the drop-down menu.</li>
      </ul></li>
    <li><strong>To view selected structures, click the <strong>"View"</strong>
        button.
    </strong><br />
      <ul>
        <li>Additional structure data will be downloaded with the
          EMBL-EBI's dbfetch service</li>
        <li><a href="siftsmapping.html">SIFTS</a> records will also
          be downloaded for mapping UniProt protein sequence data to PDB
          coordinates.</li>
        <li>A new structure viewer will open, or you will be
          prompted to add structures to existing viewers (see <a
          href="#afterviewbutton">below</a> for details).
        </li>
      </ul></li>
  </ol>
  <p>
    <strong>Structure Viewers in the Jalview Desktop</strong><br /> The
    <a href="jmol.html">Jmol viewer</a> has been included since Jalview
    2.3. Jalview 2.8.2 included support for <a href="chimera.html">Chimera</a>,
    provided it is installed and can be launched by Jalview. ChimeraX and PyMOL
    support is included from Jalview 2.11.2. The default
    viewer can be configured in the <a href="preferences.html#structure">Structure
      tab</a> in the <strong>Tools&rarr;Preferences</strong> dialog box.
  
  <p>
    Structure data imported into Jalview can also be processed to
    display secondary structure and temperature factor annotation. See
    the <a href="xsspannotation.html">Annotation from Structure</a> page
    for more information.
  </p>
  <p>
    <strong><a
      name="afterviewbutton">Controlling where the new structures
        will be shown</a></strong>
    <br/>
    <img src="schooser_viewbutton.png" />
        <br />The Structure Chooser offers several options
    for viewing a structure. <br/>
    The <strong>Open new structure view</strong> button will open a new
    structure viewer for the selected structures, but if there are views
    already open, you can select which one to use, and press the <strong>Add</strong>
    button. Jalview can automatically superimpose new structures based
    on the linked alignments - but if this is not desirable, simple
    un-tick the <strong>Superpose Structures</strong> checkbox.

  </p>
  <p>
    <em>Superposing structures</em><br/>Jalview superposes structures using
    the visible portions of any associated sequence alignments. A
    message in the structure viewer's status bar will be shown if not
    enough aligned columns were available to perform a superposition.
  </p>
  <p>
  See the <a href="jmol.html">Jmol
    </a> and <a href="chimera.html">Chimera</a> help pages for
      more information about their capabilities.</p>
  

  <p>
    <strong>Retrieving sequences from the PDB</strong><br>You can
    retrieve sequences from the PDB using the <a
      href="pdbsequencefetcher.html">Sequence Fetcher</a>. The sequences
    retrieved with this service are derived directly from the PDB 3D
    structure data, which can be viewed in the same way above. Secondary
    structure and temperature factor annotation can also be added. <br />

    <br>Jalview will also read PDB files directly - either in PDB
    format, or <a href="mmcif.html">mmCIF</a>. Simply load in the file
    as you would an alignment file. The sequences of any protein or
    nucleotide chains will be extracted from the file and viewed in the
    alignment window.
  </p>

  <p>
    <strong>Associating a large number of PDB files to
      sequences in an alignment</strong><br /> It is often the case when working
    with structure alignments that you will have a directory of PDB
    files, and an alignment involving one or more of the structures. If
    you drag a number of PDB files onto an alignment in the Jalview
    desktop, Jalview will give you the option of associating PDB files
    with sequences that have the same filename. This means, for example,
    you can automatically associate PDB files with names like '1gaq.pdb'
    with sequences that have an ID like '1gaq'. <br /> <em>Note:
      This feature was added in Jalview 2.7</em>
  </p>
  <p>
    <em>Note for Jalview applet users:<br> Due to the applet
      security constraints, PDB Files can currently only be imported by
      cut and paste of the PDB file text into the text box opened by the
      'From File' entry of the structure menu.
    </em>
  </p>

  <p>
    <strong>Viewing the PDB Residue Numbering</strong><br>
    Sequences which have PDB entry or PDB file associations are
    annotated with sequence features from a group named with the
    associated PDB accession number or file name. Each feature gives the
    corresponding PDB Residue Number for each mapped residue in the
    sequence. The display of these features is controlled through the <strong>"View&rarr;Sequence
      Features"</strong> menu item and the <a href="featuresettings.html">Feature
      Settings dialog box</a>.
  </p>
  <br />
  <hr>
  <p>
    <strong>Switching between mmCIF and PDB format for
      downloading files from the PDB</strong><br /> Jalview now employs the <a
      href="mmcif.html">mmCIF format</a> for importing 3D structure data
    from flat file and EMBL-PDBe web-service, as recommended by the
    wwwPDB. If you prefer (for any reason) to download data as PDB files
    instead, then first close Jalview, and add the following line to
    your .jalview_properties file:<br />
    <code> PDB_DOWNLOAD_FORMAT=PDB </code>
    <br /> When this setting is configured, Jalview will only request
    PDB format files from EMBL-EBI's PDBe.<br /> <em>mmCIF format
      file support was added in Jalview 2.10.</em>
  </p>

  <p>
    <em><strong>Outstanding problem with cut'n'pasted
        files in Jalview 2.6 and Jalview 2.7</strong><br>Structures imported
      via the cut'n'paste dialog box will not be correctly highlighted
      or coloured when they are displayed in structure views, especially
      if they contain more than one PDB structure. See the bug report at
      http://issues.jalview.org/browse/JAL-623 for news on this problem.</em>
  </p>

</body>
</html>

