<html>
<!--
 * Jalview - A Sequence Alignment Editor and Viewer (2.11.2.5)
 * Copyright (C) 2022 The Jalview Authors
 * 
 * This file is part of Jalview.
 * 
 * Jalview is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License 
 * as published by the Free Software Foundation, either version 3
 * of the License, or (at your option) any later version.
 *  
 * Jalview is distributed in the hope that it will be useful, but 
 * WITHOUT ANY WARRANTY; without even the implied warranty 
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR 
 * PURPOSE.  See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Jalview.  If not, see <http://www.gnu.org/licenses/>.
 * The Jalview Authors are detailed in the 'AUTHORS' file.
 -->
<head>
<title>JPred Secondary Structure Prediction</title>
</head>
<body>
  <strong>JPred Secondary Structure Prediction</strong>
  <p>
    Secondary structure prediction methods attempts to infer the likely
    secondary structure for a protein based on its amino acid
    composition and similarity to sequences with known secondary
    structure. The most recent version of the method, JPred4, employs a
    series of neural networks trained to predict different secondary
    structure types from a sequence profile, and when necessary, employs
    a jury network to identify the most likely secondary structure
    prediction.<br>
  <ul>
    <li>Drozdetskiy A, Cole C, Procter J & Barton GJ. (2015)<br />
      JPred4: a protein secondary structure prediction server<br /> <em>Nucleic
        Acids Research</em>, <strong>Web Server issue</strong> (first
      published 15th April 2015)<br /> <a
      href="http://dx.doi.org/10.1093/nar/gkv332">http://dx.doi.org/10.1093/nar/gkv332</a>
    </li>
    <li>Cole C., Barber J.D. and Barton G.J. (2008) The Jpred 3
      secondary structure prediction server <em>Nucleic Acids
        Research</em> <strong>36</strong> W197-W201
    </li>
    <li>Cuff J. A and Barton G.J (1999) Application of enhanced
      multiple sequence alignment profiles to improve protein secondary
      structure prediction <em>Proteins</em> <strong>40</strong> 502-511
    </li>
  </ul>
  </p>
  The function available from the
  <strong>Web Service&#8594;Secondary Structure
    Prediction&#8594;JPred Secondary Structure Prediction</strong> menu does two
  different kinds of prediction, dependent upon the currently selected
  region:
  </p>
  <ul>
    <li>If nothing is selected, and the displayed sequences appear
      to be aligned, then a JPred prediction will be run for the first
      sequence in the alignment, using the current alignment. Otherwise
      the first sequence will be submitted for prediction.</li>
    <li>If just one sequence (or a region on one sequence) has been
      selected, it will be submitted to the automatic JPred prediction
      server for homolog detection and prediction.</li>
    <li>If a set of sequences are selected, and they appear to be
      aligned, then the alignment will be used for a JPred prediction on
      the <strong>first</strong> sequence selected in the set (that is,
      the one nearest the top of the alignment window).
    </li>
  </ul>
  <p>
    <strong>Note</strong>: JPred secondary structure prediction is a
    'non-column-separable' service - predictions are based on the
    sequence profile of contiguous stretches of amino-acid sequence. A
    prediction will only be made on the visible parts of a sequence (see
    <a href="../features/hiddenRegions.html">hiding columns</a>) as if
    it were a contiguous polypeptide chain. Prediction accuracy at the
    hidden column boundaries may therefore be less than indicated by
    JPred's own reliability score (see below).
  </p>
  <p>The result of a JPred prediction for a sequence is a new
    annotated alignment window:</p>
  <img src="jnetprediction.gif">
  <p>The sequence for which the prediction was made is the first one
    in the alignment. If a sequence based prediction was made then the
    remaining sequences in the alignment are the aligned parts of
    homologs which were used to construct a sequence profile for the
    prediction. If the prediction was made using a multiple alignment,
    then the original multiple alignment will be returned, annotated
    with the prediction.</p>
  The annotation bars below the alignment are as follows:
  </p>
  <ul>
    <li>Lupas_21, Lupas_14, Lupas_28<br> <em>Coiled-coil
        predictions for the sequence. These are binary predictions for
        each location.</em></li>
    <li>Jnet Burial<br> <em>Prediction of Solvent
        Accessibility. levels are
        <ul>
          <li>0 - Exposed</li>
          <li>3 - 25% or more S.A. accessible</li>
          <li>6 - 5% or more S.A. accessible</li>
          <li>9 - Buried (<5% exposed)</li>
        </ul></li>
    <li>JNetPRED<br> <em>The consensus prediction -
        helices are marked as red tubes, and sheets as dark green
        arrows.</em></li>
    <li>JNetCONF<br> <em>The confidence estimate for the
        prediction. High values mean high confidence. prediction -
        helices are marked as red tubes, and sheets as dark green
        arrows.</em></li>
    <li>JNetALIGN<br> <em>Alignment based prediction -
        helices are marked as red tubes, and sheets as dark green
        arrows.</em></li>
    <li>JNetHMM<br> <em>HMM profile based prediction -
        helices are marked as red tubes, and sheets as dark green
        arrows.</em></li>
    <li>JNETPSSM<br> <em>PSSM based prediction - helices
        are marked as red tubes, and sheets as dark green arrows.</em></li>
    <li>JNETJURY<br> <em>A '*' in this annotation
        indicates that the JNETJURY was invoked to rationalise
        significantly different primary predictions.</em></li>
  </ul>
  </p>
  <em>JPred annotation created in Jalview 2.8.2 and later versions
    can be displayed on other alignments via the <a
    href="../features/annotation.html#seqannots">Add reference
      annotation</a> Sequence ID popup menu option.
  </em>
  <em>As of Jalview 2.6, the JPred service accessed accessed via
    the 'Secondary structure prediction' submenu should be considered a
    legacy Jalview SOAP service, and will be replaced in the near future
    by a JPred4 Rest service.</em>

</body>
</html>
