<html>
<!--
 * Jalview - A Sequence Alignment Editor and Viewer (2.11.2.5)
 * Copyright (C) 2022 The Jalview Authors
 * 
 * This file is part of Jalview.
 * 
 * Jalview is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License 
 * as published by the Free Software Foundation, either version 3
 * of the License, or (at your option) any later version.
 *  
 * Jalview is distributed in the hope that it will be useful, but 
 * WITHOUT ANY WARRANTY; without even the implied warranty 
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR 
 * PURPOSE.  See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Jalview.  If not, see <http://www.gnu.org/licenses/>.
 * The Jalview Authors are detailed in the 'AUTHORS' file.
 -->
<head>
<title>Alignment Conservation Annotation</title>
</head>
<body>
  <p>
    <strong>Alignment Conservation Annotation</strong>
  </p>
  <p>
    This is an automatically calculated quantitative alignment
    annotation which measures the number of conserved physico-chemical
    properties conserved for each column of the alignment. Its
    calculation is based on the one used in the AMAS method of multiple
    sequence alignment analysis :<br>
  <ul>
    Livingstone C.D. and Barton G.J. (1993), Protein Sequence
    Alignments: A Strategy for the Hierarchical Analysis of Residue
    Conservation.
    <em>CABIOS</em> Vol.
    <b>9</b> No. 6 (745-756)).
  </ul>
  <em><a
    href="http://www.compbio.dundee.ac.uk/papers/amas/amas3d.html">View
      an HTML version of the paper</a></em>
  </p>
  <p>
    Conservation is measured as a numerical index reflecting the
    conservation of <a href="../misc/aaproperties.html">physico-chemical
      properties</a> in the alignment: Identities score highest, and the
    next most conserved group contain substitutions to amino acids lying
    in the same physico-chemical class.
  </p>
  <p>Conservation is visualised on the alignment or a sequence group
    as a histogram giving the score for each column. Conserved columns
    are indicated by '*' (score of 11 with default amino acid property
    grouping), and columns with mutations where all properties are
    conserved are marked with a '+' (score of 10, indicating all
    properties are conserved).</p>
  <p>
    Mousing over a conservation histogram reveals a tooltip which
    contains a series of symbols corresponding to the physicochemical
    properties that are conserved amongst the amino acids observed at
    each position. In these tooltips, the presence of <em>!</em> implies
    that the lack of a particular physicochemical property is conserved
    (e.g. !proline).
  </p>
  <p>
    <strong>Colouring an alignment by conservation</strong><br>
    Conservation scores can be used to colour an alignment. This is
    explained further in the help page for <a
      href="../colourSchemes/conservation.html">conservation
      colouring</a>.
  </p>
  <p>
    <strong>Group conservation</strong><br> If sequence groups have
    been defined, then selecting option 'Group Conservation' in the <a
      href="../menus/alwannotation.html">Annotations menu</a> will
    result in Conservation being calculated for each group, as well as
    the alignment as a whole.
  </p>
</body>
</html>
