<html>
<!--
 * Jalview - A Sequence Alignment Editor and Viewer (2.11.2.5)
 * Copyright (C) 2022 The Jalview Authors
 * 
 * This file is part of Jalview.
 * 
 * Jalview is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License 
 * as published by the Free Software Foundation, either version 3
 * of the License, or (at your option) any later version.
 *  
 * Jalview is distributed in the hope that it will be useful, but 
 * WITHOUT ANY WARRANTY; without even the implied warranty 
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR 
 * PURPOSE.  See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Jalview.  If not, see <http://www.gnu.org/licenses/>.
 * The Jalview Authors are detailed in the 'AUTHORS' file.
 -->
<head>
<title>Alignment Window Menus</title>
</head>

<body>
  <p>
    <strong>Alignment Window Calculate Menu</strong>
  </p>
  <ul>
    <li><strong>Sort </strong>
      <ul>
        <li><strong>By ID</strong><em><br> This will sort
            the sequences according to sequence name. If the sort is
            repeated, the order of the sorted sequences will be
            inverted. </em></li>
        <li><strong>By Length</strong><em><br> This will
            sort the sequences according to their length (excluding gap
            characters). If the sort is repeated, the order of the
            sorted sequences will be inverted. </em></li>
        <li><strong>By Group</strong><strong><br> </strong><em>This
            will sort the sequences according to sequence name. If the
            sort is repeated, the order of the sorted sequences will be
            inverted. </em><strong></strong></li>
        <li><strong>By Pairwise Identity<br>
        </strong><em>This will sort the selected sequences by their
            percentage identity to the consensus sequence. The most
            similar sequence is put at the top. </em></li>
        <li><em>The <a href="../calculations/sorting.html">Sort
              menu</a> will have some additional options if the alignment
            has any associated score annotation, or you have just done a
            multiple alignment calculation or opened a tree viewer
            window.
        </em><br></li>
      </ul></li>
    <li><strong>Calculate Tree or PCA ...</strong> <br> <em>Opens the 
    <a href="../calculations/calculations.html">calculations dialog</a> for
        for calculating <a href="../calculations/tree.html">trees</a> or
         <a href="../calculations/pca.html">principal component analysis 
         plots</a> on the alignment or the currently selected
        region. 
      </em><br>
    </li>
    <li><strong>Pairwise Alignments</strong><br> <em>Applies
        Smith and Waterman algorithm to selected sequences. See <a
        href="../calculations/pairwise.html">pairwise
          alignments</a>.
    </em><br></li>
    <li><strong>Extract Scores ... (optional)</strong><br> <em>This
        option is only visible if Jalview detects one or more
        white-space separated values in the description line of the
        alignment sequences.<br> When selected, these numbers are
        parsed into sequence associated annotation which can then be
        used to sort the alignment via the Sort by&#8594;Score menu.
    </em> <br></li>
    <li><a name="transcdna"><strong>Translate as cDNA</strong></a><br> <em>This
        option is visible for nucleotide alignments. Selecting this
        option shows the DNA's calculated protein product in a new <a
        href="../features/splitView.html">split frame</a> window. Note
        that the translation is not frame- or intron-aware; it simply
        translates all codons in each sequence. You can use the standard <a
        href="../misc/geneticCode.html">genetic code</a>, or choose an 
        alternative code table (any incomplete final codon is discarded). 
        You can perform this action on the whole alignment, or selected 
        rows, columns, or regions. Alternative code tables were added in
        Jalview 2.11.
    </em> <br></li>
    <li><strong>Reverse, Reverse Complement</strong> (not applet)<br>
      <em>These options are visible for nucleotide alignments.
        Selecting them adds the reverse (or reverse complement) of the
        sequences (or selected region) as new sequences in the
        alignment. To try this out, add this sequence and perform
        'Reverse Complement' followed by 'Translate as cDNA': <br>
      <small> Seq
          GTCATTTGCGCGTGTTGATTATTCGGACCGCTCCACTTCCCTTTACTCGTGCGTTCAATTGATTTAATCCTC
          TGGGGGGGCTCTGGTTTACATAGCTTAAATCTATTCCATTCAAGGAAGCTCATG</small>
    </em> <br></li>
    <li><strong>Get Cross-References</strong> (not applet)<br>
      <em>This option is visible where sequences have
        cross-references to other standard databases; for example, an
        EMBL entry may have cross-references to one or more UNIPROT
        entries. Select the database to view all cross-referenced
        sequences in a new <a href="../features/splitView.html">split
          frame</a> window.
    </em> <br></li>
    <li><strong>Autocalculate Consensus</strong><br> <em>For
        large alignments it can be useful to deselect
        &quot;Autocalculate Consensus&quot; when editing. This prevents
        the sometimes lengthy calculations performed after each sequence
        edit.</em> <br></li>
    <li><strong>Sort Alignment With New Tree</strong><br> <em>If
        this option is selected, the alignment will be automatically
        sorted whenever a new tree is calculated or loaded.</em> <br></li>
    <li><strong>Show Flanking Regions</strong><br> <em>Opens
        a new alignment window showing any additional sequence data
        either side of the current alignment. Useful in conjunction with
        'Fetch Database References' when the 'Trim Retrieved Sequences'
        option is disabled to retrieve full length sequences for a set
        of aligned peptides. </em></li>
  </ul>
</body>
</html>
